fish-rust/src/abbrs.rs

```diff
@@ -0,0 +1,442 @@
+        set_cursor_marker: UniquePtr<CxxWString>,
+        has_cursor_marker: bool,
+    }
+
+    struct abbrs_replacement_t {
+        range: SourceRange,
+        text: UniquePtr<CxxWString>,
+        cursor: usize,
+        has_cursor: bool,
+    }
+
+    struct abbreviation_t {
+        key: UniquePtr<CxxWString>,
+        replacement: UniquePtr<CxxWString>,
+        is_regex: bool,
+    }
+
+    extern "Rust" {
+        type GlobalAbbrs<'a>;
+
+        #[cxx_name = "abbrs_list"]
+        fn abbrs_list_ffi() -> Vec<abbreviation_t>;
+
+        #[cxx_name = "abbrs_match"]
+        fn abbrs_match_ffi(token: &CxxWString, position: abbrs_position_t)
+            -> Vec<abbrs_replacer_t>;
+
+        #[cxx_name = "abbrs_has_match"]
+        fn abbrs_has_match_ffi(token: &CxxWString, position: abbrs_position_t) -> bool;
+
+        #[cxx_name = "abbrs_replacement_from"]
+        fn abbrs_replacement_from_ffi(
+            range: SourceRange,
+            text: &CxxWString,
+            set_cursor_marker: &CxxWString,
+            has_cursor_marker: bool,
+        ) -> abbrs_replacement_t;
+
+        #[cxx_name = "abbrs_get_set"]
+        unsafe fn abbrs_get_set_ffi<'a>() -> Box<GlobalAbbrs<'a>>;
+        unsafe fn add<'a>(
+            self: &mut GlobalAbbrs<'_>,
+            name: &CxxWString,
+            key: &CxxWString,
+            replacement: &CxxWString,
+            position: abbrs_position_t,
+            from_universal: bool,
+        );
+    }
+}
+
+static abbrs: Lazy<Arc<Mutex<AbbreviationSet>>> =
+    Lazy::new(|| Arc::new(Mutex::new(Default::default())));
+
+pub fn with_abbrs<R>(cb: impl FnOnce(&AbbreviationSet) -> R) -> R {
+    let abbrs_g = abbrs.lock().unwrap();
+    cb(&abbrs_g)
+}
+
+pub fn with_abbrs_mut<R>(cb: impl FnOnce(&mut AbbreviationSet) -> R) -> R {
+    let mut abbrs_g = abbrs.lock().unwrap();
+    cb(&mut abbrs_g)
+}
+
+/// Controls where in the command line abbreviations may expand.
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Position {
+    Command,  // expand in command position
+    Anywhere, // expand in any token
+}
+
+impl From<abbrs_position_t> for Position {
+    fn from(value: abbrs_position_t) -> Self {
+        match value {
+            abbrs_position_t::anywhere => Position::Anywhere,
+            abbrs_position_t::command => Position::Command,
+            _ => panic!("invalid abbrs_position_t"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Abbreviation {
+    // Abbreviation name. This is unique within the abbreviation set.
+    // This is used as the token to match unless we have a regex.
+    pub name: WString,
+
+    /// The key (recognized token) - either a literal or a regex pattern.
+    pub key: WString,
+
+    /// If set, use this regex to recognize tokens.
+    /// If unset, the key is to be interpreted literally.
+    /// Note that the fish interface enforces that regexes match the entire token;
+    /// we accomplish this by surrounding the regex in ^ and $.
```

Comment:
yeah`regex: Option<regex_t>,` will not work but you can make it a `UniquePtr<regex_t>.
I think that works. If we run into problems we can port `re.h` first, it has no dependencies.